src/core/DY.Core.Trace.Base.fst

```diff
@@ -367,29 +465,96 @@ let rec fmap_trace_compose #a #b #c f g h tr =
+val fmap_trace_prefix:
+  #a:Type -> #b:Type ->
+  f:(a -> b) ->
+  tr:trace_ a -> i:timestamp{i <= length tr} ->
+  Lemma (
+    prefix (fmap_trace f tr) i == fmap_trace f (prefix tr i)
+  )
+let rec fmap_trace_prefix #a #b f tr i =
+  norm_spec [zeta; delta_only [`%prefix]] (prefix #a);
+  norm_spec [zeta; delta_only [`%prefix]] (prefix #b);
+  if length tr = i then ()
+  else
+    let Snoc tr_init _ = tr in
+    fmap_trace_prefix f tr_init i
 
-val trace_forget_labels_later:
-  tr1:trace -> tr2:trace ->
+val fmap_trace_later:
+  #a:Type -> #b:Type ->
+  f:(a -> b) ->
+  tr1:trace_ a -> tr2:trace_ a ->
   Lemma
   (requires tr1 <$ tr2)
-  (ensures (trace_forget_labels tr1) <$ (trace_forget_labels tr2))
-let rec trace_forget_labels_later tr1 tr2 =
-  reveal_opaque (`%grows) (grows #unit);
-  norm_spec [zeta; delta_only [`%prefix]] (prefix #unit);
-  reveal_opaque (`%grows) (grows #label);
-  norm_spec [zeta; delta_only [`%prefix]] (prefix #label);
-  if length tr1 = length tr2 then (
-    match tr1, tr2 with
-    | Nil, Nil -> ()
-    | Snoc tr1_init _, Snoc tr2_init _ -> ()
-  ) else (
-    let Snoc tr2_init tr2_last = tr2 in
-    trace_forget_labels_later tr1 tr2_init
+  (ensures (fmap_trace f tr1) <$ (fmap_trace f tr2))
+let fmap_trace_later #a #b f tr1 tr2 =
+  reveal_opaque (`%grows) (grows #a);
+  reveal_opaque (`%grows) (grows #b);
+  fmap_trace_prefix f tr2 (length tr1)
+
+val fmap_trace_recover_before:
+  #a:Type -> #b:Type ->
+  f:(a -> b) ->
+  tr1:trace_ b -> tr2:trace_ a ->
+  Pure (trace_ a)
+  (requires
+    tr1 <$ fmap_trace f tr2
+  )
+  (ensures fun tr1' ->
+    tr1 == fmap_trace f tr1' /\
+    tr1' <$ tr2
+  )
+let fmap_trace_recover_before #a #b f tr1 tr2 =
+  reveal_opaque (`%grows) (grows #b);
+  fmap_trace_prefix f tr2 (length tr1);
+  prefix tr2 (length tr1)
+
+val get_event_at_fmap_trace:
+  #a:Type -> #b:Type ->
+  f:(a -> b) -> tr:trace_ a -> i:timestamp{i < length tr} ->
+  Lemma (
+    get_event_at (fmap_trace f tr) i == fmap_trace_event f (get_event_at tr i)
+  )
+let rec get_event_at_fmap_trace #a #b f tr i =
+  if i+1 = length tr then ()
+  else (
+    let Snoc tr_init _ = tr in
+    get_event_at_fmap_trace f tr_init i
+  )
+
+val event_at_fmap_trace:
+  #a:Type -> #b:Type ->
+  f:(a -> b) -> tr:trace_ a -> i:timestamp -> ev:trace_event_ a ->
+  Lemma
+  (requires event_at tr i ev)
+  (ensures event_at (fmap_trace f tr) i (fmap_trace_event f ev))
+let event_at_fmap_trace #a #b f tr i ev =
+  if i >= length tr then ()
+  else (
+    get_event_at_fmap_trace f tr i
+  )
+
+val event_exists_fmap_trace:
+  #a:Type -> #b:Type ->
+  f:(a -> b) -> tr:trace_ a -> ev:trace_event_ a ->
+  Lemma
+  (requires event_exists tr ev)
+  (ensures event_exists (fmap_trace f tr) (fmap_trace_event f ev))
+let event_exists_fmap_trace #a #b f tr ev =
+  eliminate exists i. event_at tr i ev
+  returns event_exists (fmap_trace f tr) (fmap_trace_event f ev)
+  with _. (
+    event_at_fmap_trace f tr i ev
   )
+
+val forget_label:
```

Comment:
Is there a benefit to leaving this so generic? It seems like this replaces all labels with a constant (the first non-implicit argument), and I don't really see an application for this other than with () as that argument --- maybe something to do with replacing all labels with `public`?

Reply:
Yes, this is used here https://github.com/REPROSEC/dolev-yao-star-extrinsic/blob/4fe16e83dc70626200966322010cb7ad2fadf09a/src/core/DY.Core.Label.fst#L136
to justify that `trace_forget_labels` is surjective (because `trace_forget_labels (fmap_trace (forget_label unknown_label) tr) == tr`)
But the it is the only place where we use `forget_label` in all its generality so it may be relevant to define
```fstar
// the old `forget_label`
val replace_label:
  #a:Type -> #b:Type ->
  b ->
  a -> b
let replace_label #a #b x _ = x

val forget_label:
  #a:Type ->
  a -> unit
let forget_label = replace_label ()
```

Reply:
Yes, this is used here https://github.com/REPROSEC/dolev-yao-star-extrinsic/blob/4fe16e83dc70626200966322010cb7ad2fadf09a/src/core/DY.Core.Label.fst#L136
to justify that `trace_forget_labels` is surjective (because `trace_forget_labels (fmap_trace (forget_label unknown_label) tr) == tr`)
But the it is the only place where we use `forget_label` in all its generality so it may be relevant to define
```fstar
// the old `forget_label`
val replace_label:
  #a:Type -> #b:Type ->
  b ->
  a -> b
let replace_label #a #b x _ = x

val forget_label:
  #a:Type ->
  a -> unit
let forget_label = replace_label ()
```

Reply:
Done in #75